Add triggered_by field to DAG Run model to distinguish the source of a trigger

molcay opened the pull request:
Add the `triggered_by` field to the DAG Run model to distinguish the source of a trigger.
This is for the discussion in https://github.com/apache/airflow/pull/37087.

> Currently this PR contains only the backend changes. No changes done on the frontend. We can discuss and add some UI elements (different border types, colors, or icons etc.) to have a more distinct look on the UI as well. 

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

molcay committed:
Add triggered_by field to DAG Run model to distinguish the source of a trigger

Taragolis commented on airflow/api/client/api_client.py:
`airflow/api/client/*` files exists for mainly to support legacy experimental API, I don't think we need to extend those

molcay replied:
Since they currently exist, I thought that I need to extend those files also. I can revert those changes.

molcay replied:
I reverted changes for the `airflow/api/client` folder. Since some of the clients are using other methods, I only touched on fixing the related places

potiuk replied:
LGTM

uranusjr commented on airflow/api/client/api_client.py:
```diff
@@ -31,14 +36,23 @@ def __init__(self, api_base_url, auth=None, session: httpx.Client | None = None)
+        conf=None,
+        execution_date=None,
+        replace_microseconds=True,
+        triggered_by: DagRunTriggeredByType | None = None,
```
What does a `None` value mean? I feel we should make this argument required instead.

molcay replied:
Actually, my initial thought was setting None here as default but setting the default value for this property has a different value like `REST_API` in the child classes. But it also makes sense to convert this parameter to a required one (I even forget to set the default values in child classes)

Also, is it OK to have this change or I need to follow @Taragolis's recommendation and revert the changes in this folder: `airflow/api/client/*`; WDYT?

uranusjr replied:
Let’s revert the changes in `api/client` first.

uranusjr commented on airflow/api/common/trigger_dag.py:
```diff
@@ -39,6 +39,7 @@ def _trigger_dag(
     conf: dict | str | None = None,
     execution_date: datetime | None = None,
     replace_microseconds: bool = True,
+    triggered_by: DagRunTriggeredByType | None = None,
```
```suggestion
    triggered_by: DagRunTriggeredByType,
```

Same as mentioned previously, I think we should make this strictly required, and keep the None case only for legacy records. There are other arguments in the call chain that should be changed as well.

molcay replied:
Your suggestion makes total sense. I will try to change the code (here and the other places) accordingly

uranusjr commented on airflow/utils/types.py:
```diff
@@ -72,3 +72,18 @@ class EdgeInfoType(TypedDict):
+class DagRunTriggeredByType(str, enum.Enum):
+    """Class with TriggeredBy types for DagRun."""
+
+    CLI = "cli"  # for the trigger subcommand for dag command in cli: airflow dags trigger
```
This does not read like English

molcay replied:
I changed the comment wording

uranusjr commented on airflow/utils/types.py:
```diff
@@ -72,3 +72,18 @@ class EdgeInfoType(TypedDict):
+    REST_API = "rest_api"  # for triggering the DAG via RESTful API
+    UI = "ui"  # for clicking the `Trigger DAG` button
+    TEST = "test"  # for dag.test()
+    SCHEDULER = "scheduler"  # for scheduler
```
This should probably be say `TIME` or `TIMETABLE` or something to distinguish from `DATASET`. Maybe something like `TIME_TRIGGERD` and `EVENT_TRIGGERED` would work too.

molcay replied:
Actually, I was thinking that the `SCHEDULER` value is for timetable related triggering and `DATASET` value is for Dataset related triggering. 
I saw [this image](https://airflow.apache.org/docs/apache-airflow/stable/_images/dataset-scheduled-dags.png) on [the documentation](https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/datasets.html) and it states "Triggered by Datasets" so I was thinking the `DATASET` is a good value for this.

potiuk replied:
@uranusjr ?

uranusjr replied:
Your interpretation is not wrong, but the problem is that both timetable-related and database-related triggering are ultimately done by the scheduler, so this name is confusing. Some parts of Airflow use “scheduler” to refer to (only) timetable-related scheduling because historically that was the only way to schedule things; datasets were introduced later. We don’t need to follow the naming. Even if we keep the `SCHEDULER` variable—I don’t think we should—a better comment is still needed to explain this does not cover dataset-triggered runs.

molcay replied:
Hi @uranusjr, I missed this comment, sorry for that. 
How about changing the line with the following
```python
    TIMETABLE = "timetable"  # for timetable scheduler
```
as you suggested at the beginning. I think this will be clear enough to distinguish timetable based triggering and dataset based triggering. WDYT?

uranusjr replied:
I think that’s fine. Except _timetable scheduler_ is awkward; your wording in the next sentence is much better.

molcay replied:
yes, you are right. How about:
```python
    TIMETABLE = "timetable"  # for timetable based triggering
```
>

uranusjr commented on airflow/api/common/trigger_dag.py:
```diff
@@ -104,6 +107,7 @@ def _trigger_dag(
 
 def trigger_dag(
     dag_id: str,
+    triggered_by: DagRunTriggeredByType,
```
This should probably be a keyword-only argument instead. Sticking a positional parameter in the middle is too easy to get wrong. Same applies for other functions that add this argument.

molcay replied:
OK, it makes sense. Do you think that it is wise to use `*` as a parameter before the `triggered_by` parameter to force to use named arguments (as below)? 
```python
def trigger_dag(
    dag_id: str,
    *,  # <- new line
    triggered_by: DagRunTriggeredByType,
```

I think the other option is setting to `None` as the default value and checking the value is `None` in the function body.  This option will definitely look ugly.

WDYT? @uranusjr

uranusjr replied:
Adding the star is the only way. Simply adding a default does not even make the argument *keyword-only*.

jscheffl reviewed (commented):
I was taking a look to this PR and see that the field will add valuable information. But not completely as besides adding the marker as type we still would not know "which CLI", which "User" in the UI or which DAG run/Operator triggered a DAG run. So it adds parts of traceability with a lot of changes in the code (1000+LoC code changed).

Have you considered adding the details for your needs as traceability to the LOG / audit log table? This would just be another type in the audit logs and would prevent the need to extend the DB scheme for the DAG run table.

potiuk reviewed (commented):
I re-reviewed it, and it looks GOOD. @uranusjr  -> one more comment to address about the names of SCHEDULED/DATASET . Does it look good for  you.

uranusjr commented on airflow/models/dagrun.py:
```diff
@@ -131,6 +131,7 @@ class DagRun(Base, LoggingMixin):
     creating_job_id = Column(Integer)
     external_trigger = Column(Boolean, default=True)
     run_type = Column(String(50), nullable=False)
+    triggered_by = Column(String(50))  # who triggered the DAG Run
```
```suggestion
    triggered_by = Column(String(50))  # Airflow component that triggered the run.
```

uranusjr commented on airflow/utils/types.py:
```diff
@@ -99,3 +99,18 @@ class EdgeInfoType(TypedDict):
     label: str | None
+
+
+class DagRunTriggeredByType(str, enum.Enum):
```
```suggestion
class DagRunTriggeredByType(enum.Enum):
```

Existing enums subclass from str due to compatibility; we don’t need that here.

molcay replied:
If we remove the `str` then the db value will be in the upper case format like: `UI`, `TIMETABLE` not `ui`, `timetable`.
Do you think that this is OK, @uranusjr ?

uranusjr replied:
That’s fine, nobody should care about those values directly anyway.

uranusjr commented on airflow/utils/types.py:
```diff
@@ -99,3 +99,18 @@ class EdgeInfoType(TypedDict):
+    DATASET = "dataset"  # for dataset_triggered run type
+
+    def __str__(self):
+        return self.value
```
```suggestion
```

Same. I think the API needs some changes to accomodate this change.

uranusjr commented on airflow/api_connexion/schemas/dag_run_schema.py:
```diff
@@ -74,6 +74,7 @@ class Meta:
     last_scheduling_decision = auto_field(dump_only=True)
     run_type = auto_field(dump_only=True)
     note = auto_field(dump_only=False)
+    triggered_by = auto_field(dump_only=True)
```
Not sure how this field would work if the enum does not subclass from str. We’ll find out with tests.

molcay replied:
It is not working with this way, I changed the line with this:
```python
    triggered_by = fields.Enum(DagRunTriggeredByType, by_value=True, dump_only=True)
```
I am trying to run the tests and see everything is fine or not

potiuk committed:
Merge branch 'main' into feat/add-manual-triggerer-type

molcay committed:
Fix conflict
Fix import statement for listener test
Fix missing parameter for triggered_by for new test case
